mypy/fastparse.py

```diff
@@ -776,7 +776,8 @@ def visit_Str(self, n: ast35.Str) -> Union[UnicodeExpr, StrExpr]:
         def visit_JoinedStr(self, n: ast35.JoinedStr) -> StrExpr:
             result_string_expression = StrExpr('')
             for value in n.values:
-                result_string_expression = OpExpr('+', result_string_expression, self.visit(value))
+                value_as_string_expr = cast(StrExpr, self.visit(value))
+                result_string_expression = cast(StrExpr, OpExpr('+', result_string_expression, value_as_string_expr))
```

Comment:
I wonder if you could use asserts instead of casts, e.g.
```python
value_as_string_expr = self.visit(value)
assert isinstance(value_as_string_expr, StrExpr)
r = OpExpr('+', result_string_expression, value_as_string_expr)
assert isinstance(r, StrExpr)
result_string_expression = r
```
That way if the type is ever not what you expect it'll fail right there rather than mysteriously failing at some later point.